buildSrc/src/main/kotlin/com/datadog/gradle/plugin/jsonschema/generator/MultiClassGenerator.kt

```diff
@@ -0,0 +1,135 @@
+            docBuilder.add(definition.description)
+            docBuilder.add("\n")
+            typeBuilder.addKdoc(docBuilder.build())
+        }
+
+        definition.options.forEach {
+            val updatedType = when (it) {
+                is TypeDefinition.Class -> it.copy(parentType = definition)
+                else -> TODO("Can't have type $it as child of a `one_of` block")
+            }
+            nestedTypes.add(updatedType)
+        }
+
+        typeBuilder.addFunction(generateMultiClassSerializer())
+
+        typeBuilder.addType(generateCompanionObject(definition, rootTypeName))
+
+        return typeBuilder
+    }
+
+    // endregion
+
+    // region Internal
+
+    private fun generateMultiClassSerializer(): FunSpec {
+        return FunSpec.builder(Identifier.FUN_TO_JSON)
+            .addModifiers(KModifier.ABSTRACT)
+            .returns(ClassNameRef.JsonElement).build()
+    }
+
+    private fun generateCompanionObject(
+        definition: TypeDefinition.MultiClass,
+        rootTypeName: String
+    ): TypeSpec {
+        return TypeSpec.companionObjectBuilder()
+            .addFunction(generateMultiClassDeserializer(definition, rootTypeName))
+            .build()
+    }
+
+    private fun generateMultiClassDeserializer(
+        definition: TypeDefinition.MultiClass,
+        rootTypeName: String
+    ): FunSpec {
+        val returnType = definition.asKotlinTypeName(rootTypeName)
+        val funBuilder = FunSpec.builder(Identifier.FUN_FROM_JSON)
+            .addAnnotation(AnnotationSpec.builder(JvmStatic::class).build())
+            .throws(ClassNameRef.JsonParseException)
+            .addParameter(Identifier.PARAM_JSON_STR, String::class)
+            .returns(returnType)
+
+        // create error variable
+        funBuilder.addStatement("val errors = mutableListOf<Throwable>()")
+
+        // try to parse against all possible types
+        val options = mutableListOf<String>()
+        definition.options.forEach {
+            val typeName = it.asKotlinTypeName(rootTypeName)
+            val variableName = "as${it.name()}"
+            funBuilder.beginControlFlow("val %L = try", variableName)
+            funBuilder.addStatement(
+                "%T.%L(%L)",
+                typeName,
+                Identifier.FUN_FROM_JSON,
+                Identifier.PARAM_JSON_STR
+            )
+            funBuilder.nextControlFlow(
+                "catch (%L: %T)",
+                Identifier.CAUGHT_EXCEPTION,
+                ClassNameRef.JsonParseException
+            )
+            funBuilder.addStatement(
+                "errors.add(%L)",
+                Identifier.CAUGHT_EXCEPTION
+            )
+            funBuilder.addStatement("null")
+            funBuilder.endControlFlow()
+            options.add(variableName)
+        }
+
+        funBuilder.addStatement(
+            "val result = arrayOf(${options.joinToString(", ")}).firstOrNull { it != null }"
```

Comment:
options.joinToString(",\n"), otherwise when you use have a long array it will look ugly.

Reply:
On my end if I have more than 3 sealed classes there it goes to another line and it breaks the statement when it generates the code. Either we do what I suggested or we find a way to increase Kotlin Poet line length ?